README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dialr <a href='https://socialresearchcentre.github.io/dialr'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/dialr)](https://cran.r-project.org/package=dialr)
[![Travis build
status](https://travis-ci.org/socialresearchcentre/dialr.svg?branch=master)](https://travis-ci.org/socialresearchcentre/dialr)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/socialresearchcentre/dialr?branch=master&svg=true)](https://ci.appveyor.com/project/gorcha/dialr)
[![Coverage
status](https://codecov.io/gh/socialresearchcentre/dialr/branch/master/graph/badge.svg)](https://codecov.io/github/socialresearchcentre/dialr?branch=master)
<!-- badges: end -->

## Overview

dialr is an R interface to [Google’s libphonenumber
library](https://github.com/googlei18n/libphonenumber). It uses the java
implementation of libphonenumber via rJava for all phone number
processing.

For a full rundown of libphonenumber see their
[GitHub](https://github.com/googlei18n/libphonenumber) and
[javadocs](https://javadoc.io/doc/com.googlecode.libphonenumber/libphonenumber/).

## Installation

You can install the released version of dialr from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("dialr")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("socialresearchcentre/dialr")
```

## Usage

``` r
library(dialr)

# Parse a character phone number vector
x <- c(0, 0123, "0404 753 123", "61410123817", "+12015550123")
x <- phone(x, "AU")

is_parsed(x)    # Was the phone number successfully parsed?
#> [1] FALSE  TRUE  TRUE  TRUE  TRUE
is_valid(x)     # Is the phone number valid?
#> [1] FALSE FALSE  TRUE  TRUE  TRUE
is_possible(x)  # Is the phone number possible?
#> [1] FALSE FALSE  TRUE  TRUE  TRUE
get_region(x)   # What region (ISO country code) is the phone number from?
#> [1] NA   NA   "AU" "AU" "US"
get_type(x)     # Is the phone number a fixed line, mobile etc.
#> [1] NA                     "UNKNOWN"              "MOBILE"              
#> [4] "MOBILE"               "FIXED_LINE_OR_MOBILE"
format(x)
#> [1] NA             "+61123"       "+61404753123" "+61410123817"
#> [5] "+12015550123"
format(x, home = "AU")
#> [1] NA                "123"             "0404753123"      "0410123817"     
#> [5] "001112015550123"

# Use with dplyr
library(dplyr)

y <- tibble(id = 1:4,
            phone1 = c(0, 0123, "0404 753 123", "61410123817"),
            phone2 = c("03 9388 1234", 1234, "+12015550123", 0),
            country = c("AU", "AU", "AU", "AU"))

y %>%
  mutate_at(vars(matches("^phone")), ~phone(., country)) %>%
  mutate_at(vars(matches("^phone")),
            list(valid = is_valid,
                 region = get_region,
                 type = get_type,
                 clean = format))
#> # A tibble: 4 x 12
#>      id       phone1       phone2 country phone1_valid phone2_valid
#>   <int>      <phone>      <phone> <chr>   <lgl>        <lgl>       
#> 1     1           NA +61393881234 AU      FALSE        TRUE        
#> 2     2       +61123      +611234 AU      FALSE        FALSE       
#> 3     3 +61404753123 +12015550123 AU      TRUE         TRUE        
#> 4     4 +61410123817           NA AU      TRUE         FALSE       
#> # … with 6 more variables: phone1_region <chr>, phone2_region <chr>,
#> #   phone1_type <chr>, phone2_type <chr>, phone1_clean <chr>,
#> #   phone2_clean <chr>
```


NEWS.md
# dialr 0.3.0

## New functions

* Retrieve carrier name with `get_carrier()` (#6).

* Add geocoding functions `get_geocode()` and `get_timezone()` (#5).

## Minor changes

* `get_example()` returns an empty phone number if the provided type is invalid for the provided region (#8).

* `phone()` constructor now supports parsing phone numbers in international format without a default region by specifying `NA` or `""` in the `region` argument (#4).

* `get_type()` gets a `strict` argument. If `TRUE`, invalid phone numbers return `NA` instead of `"UNKNOWN"`.

* Rename "One shot methods" vignette to remove redundant "dialr".

# dialr 0.2.1

## Performance improvements

Code calling libphonenumber methods via rJava has been tweaked for performance,
including using more efficient rJava acess methods and caching commonly used
static values. Most operations are now approx. 10 times faster.

## Major changes

* Updated R dependency to R >= 3.2.3. R 3.6.0 just came out and highr, a dependency of knitr, requires >= 3.2.3 so build fails on 3.1.x.

## Minor changes

* New introductory `vignette("dialr")`.

* New `is_match()` function supports proper `phone` equality checks. `==` and
`!=` now use `is_match()` for `phone` vectors. (#1)

* More region and type functions implemented. See `` ?`dialr-region` `` or
`` ?`dialr-type` `` for details.

* New `get_example()` function generates example phone numbers for provided
regions and types.

* Improved function documentation.

# dialr 0.2.0

## Introduction of `dialrjars`

To separate `libphonenumber` updates from structural package updates, the
`libphonenumber` jars have been split into the new package `dialrjars`.
This is a breaking change, and `dialrjars` will need to be installed for
`dialr` to continue to work.

## Addition of the phone class

In addition to the one-shot phone processing functions, dialr now has a
"phone" S3 class to reduce the processing load from re-parsing phone numbers for
every operation. This is a fundamental package change with a full set of
replacement functions, for details see `?phone`.

# dialr 0.1.0

Initial release.


R/phone.R
#' Phone number parsing and formatting
#'
#' A phone vector stores phone numbers parsed with libphonenumber for formatting
#' and further processing.
#'
#' libphonenumber defines the `PhoneNumberUtil` class, with a set of functions
#' for extracting information from and performing processing on a parsed
#' `Phonenumber` object. A text phone number must be parsed before any other
#' operations (e.g. checking phone number validity, formatting) can be
#' performed. When parsing a phone number a ["default region"][dialr-region] is
#' required to determine the processing context for non-international numbers.
#'
#' A phone vector stores the raw phone number, the default region and a java
#' `Phonenumber` object for each element. The java object is cached so should
#' persist between R sessions. In case of issues, use `phone_reparse()` to
#' recreate the phone vector from the original phone number and region.
#'
#' Phone number parsing functions display a progress bar in interactive sessions
#' by default. This can be disabled by setting option `dialr.show_progress` to
#' `FALSE`.
#'
#' @section libphonenumber reference:
#'
#'   `phone()`: Phone numbers are parsed using
#'   `PhoneNumberUtil.parseAndKeepRawInput()`. A phone vector stores the
#'   returned `Phonenumber.PhoneNumber` object alongside the original raw text
#'   and default region for later reference.
#'
#'   `format()`: `PhoneNumberUtil.format()` by default, or
#'   `PhoneNumberUtil.formatOutOfCountryCallingNumber()` if `home` is provided.
#'   
#' @param x A character vector of phone numbers.
#' @param region A character vector of [ISO country codes][dialr-region] with
#'   the default region for each phone number in `x`. A `region` vector of
#'   length 1 will be recycled to the length of `x`.
#'   
#'   If `NA` or `""`, numbers written in international format (with a leading
#'   `+`) will be parsed without a default region.
#' @examples
#'   # Create a phone vector
#'   x <- phone(c(0, 0123, "0412 345 678", "61412987654", "03 9123 4567", "+12015550123"), "AU")
#'   
#'   is.phone(x)
#'   print(x)
#'   as.character(x)
#'   format(x)
#'   format(x, home = "AU")
#'   
#'   # Parse international number with no default region
#'   phone("+61412345678", NA)
#'   
#'   # Will fail to parse if number is not in international format
#'   phone("0412345678", NA)
#'   
#'   # A combination can be used
#'   phone(c("+61412345678", "0412345678"), c(NA, "AU"))
#' @name dialr-phone
#' @family phone functions
#' @export
phone <- function(x, region) {
  if (!is.atomic(x))  stop("`x` must be an atomic vector.", call. = FALSE)
  if (length(x) == 0)  stop("`x` must not be empty.", call. = FALSE)
  if (length(x) > 1 & length(region) == 1) region <- rep(region, length(x))
  if (length(x) != length(region)) stop("`x` and `region` vectors must be the same length.", call. = FALSE)
  
  region[!is.na(region) & region == ""] <- NA_character_
  validate_phone_region(region[!is.na(region)])

  x <- as.character(x)
  validate_phone(new_phone(x, region))
}

#' @importFrom utils txtProgressBar
#' @importFrom utils getTxtProgressBar
#' @importFrom utils setTxtProgressBar
new_phone <- function(x, region) {
  stopifnot(is.character(x))
  stopifnot(is.character(region))
  stopifnot(length(x) == length(region))

  phone_util <- .get_phoneNumberUtil()
  jfunc <- function(p, r) {
    .jcall(phone_util,
           "Lcom/google/i18n/phonenumbers/Phonenumber$PhoneNumber;",
           "parseAndKeepRawInput",
           .jstr_to_char(p),
           r)
  }
  
  show_pb <- isTRUE(getOption("dialr.show_progress")) && interactive()
  
  if (show_pb) pb <- txtProgressBar(min = 0, max = length(x), style = 3)
  out <- structure(
    mapply(
      function(p, r) {
        if (show_pb) setTxtProgressBar(pb, getTxtProgressBar(pb) + 1)
        pn <- tryCatch({
          jfunc(p, r)
        }, error = function(e) {
          return(NULL)
        })
        if (is.null(pn))
          pn <- NA
        else
          .jcache(pn)
        
        list(raw = p,
             region = r,
             jobj = pn)
      },
      x, region,
      SIMPLIFY = FALSE
    ),
    class = "phone"
  )
  if (show_pb) close(pb)
  
  names(out) <- NULL
  out
}

validate_phone <- function(x) {
  if (!inherits(x, "phone"))
    stop("`x` must be a vector of class `phone`.", call. = FALSE)
  
  x_raw <- unclass(x)
  
  if ((!is.list(x_raw)) | (!all(vapply(x_raw, is.list, logical(1)))))
    stop("`x` must be a list of lists.", call. = FALSE)

  # check structure
  if (!(all(vapply(x_raw, length, integer(1)) == 3L) &
        all(vapply(x_raw, function(x) { exists("raw", x, mode = "character", inherits = FALSE) }, logical(1))) &
        all(vapply(x_raw, function(x) { exists("region", x, mode = "character", inherits = FALSE) }, logical(1))) &
        all(vapply(x_raw, function(x) { exists("jobj", x, inherits = FALSE) }, logical(1))))) {
    stop(
      "The structure of `x` is incorrect.\n",
      "`x` should be a list.\n",
      "All elements of `x` should be named lists with 3 elements:\n",
      " * raw: a character vector of length 1\n",
      " * region: a character vector of length 1\n",
      " * jobj: an S4 `jobjRef` or `NULL`",
      call. = FALSE
    )
  }

  x
}

#' @rdname dialr-phone
#' @export
phone_reparse <- function(x) {
  if (!is.phone(x)) stop("`x` must be a vector of class `phone`.", call. = FALSE)
  
  x <- unclass(x)
  new_phone(vapply(x, function(x) { x$raw }, character(1), USE.NAMES = FALSE),
            vapply(x, function(x) { x$region }, character(1), USE.NAMES = FALSE))
}

#' @rdname dialr-phone
#' @export
is.phone <- function(x) inherits(x, "phone")

#' @export
`[.phone` <- function(x, ...) {
  structure(NextMethod(), class = "phone")
}

#' @export
`[[.phone` <- function(x, ...) {
  `[`(x, ...)
}

#' @export
`$.phone` <- function(x, ...) {
  stop("$ operator is invalid for `phone` vectors.", call. = FALSE)
}

#' @export
`[<-.phone` <- function(x, i, value) {
  if (!is.phone(value) & is.atomic(value)) {
    warning("Only `phone` class objects can be inserted into a `phone` vector.\n",
            "The value provided will be converted to `phone` class with default home region `", getOption("dialr.home"), "`.",
            call. = FALSE)
    value <- new_phone(as.character(value),
                       rep(getOption("dialr.home"), length(value)))
  } else if (!is.phone(value) & !is.atomic(value)) {
    stop("Only `phone` class objects can be inserted into a `phone` vector.\n",
         "The value provided can not be converted to `phone` class.",
         call. = FALSE)
  }
  
  NextMethod()
}

#' @export
`[[<-.phone` <- function(x, i, value) {
  `[<-`(x, i, value)
}

#' @export
`$<-.phone` <- function(x, i, value) {
  stop("$ operator is invalid for `phone` vectors.", call. = FALSE)
}

#' @export
`length<-.phone` <- function(x, value) {
  structure(NextMethod(), class = "phone")
}

#' @export
c.phone <- function(..., recursive = FALSE) {
  out <- lapply(list(...), function(value) {
    if (!is.phone(value)) {
      warning("Only `phone` class objects can be added to a `phone` vector.\n",
              "Atomic vectors will be converted to `phone` class with default home region `", getOption("dialr.home"), "`.\n",
              "Other objects will be dropped.",
              call. = FALSE)
      
      if (is.atomic(value))
        value <- new_phone(as.character(value),
                           rep(getOption("dialr.home"), length(value)))
      else
        value <- NULL
      
      value
    }
    value
  })
  
  structure(unlist(c(lapply(out, unclass)), recursive = FALSE), class = "phone")
}

#' @export
rep.phone <- function(x, ...) {
  structure(NextMethod(), class = "phone")
}

#' @rdname dialr-phone
#' @param n Number of elements to print.
#' @param ...	Additional arguments for specific methods.
#' @export
print.phone <- function(x, n = 10, ...) {
  tot <- length(x)

  cat("# Parsed phone numbers: ",
      tot, " total, ",
      sum(is_parsed(x)), " successfully parsed",
      sep = "")

  x_raw <- vapply(unclass(x), function(x) { x$raw }, character(1), USE.NAMES = FALSE)
  if (tot > n) {
    cat(" (showing first ", n, ")\n", sep = "")
    print.default(head(x_raw, n = n), quote = FALSE)
  } else {
    cat("\n")
    print.default(x_raw, quote = FALSE)
  }
  
  invisible(x)
}

#' @rdname dialr-phone
#' @param format Phone number format to use from one of four standards:
#'
#'   * `"E164"`: general format for international telephone numbers from [ITU-T
#'   Recommendation E.164](https://en.wikipedia.org/wiki/E.164)
#'
#'   * `"NATIONAL"`: national notation from [ITU-T Recommendation
#'   E.123](https://en.wikipedia.org/wiki/E.123)
#'
#'   * `"INTERNATIONAL"`: international notation from [ITU-T Recommendation
#'   E.123](https://en.wikipedia.org/wiki/E.123)
#'
#'   * `"RFC3966"`: "tel" URI syntax from the IETF [tel URI for Telephone
#'   Numbers](https://datatracker.ietf.org/doc/rfc3966/)
#'   
#'   See notes from the [libphonenumber
#'   javadocs](https://static.javadoc.io/com.googlecode.libphonenumber/libphonenumber/8.10.10/index.html?com/google/i18n/phonenumbers/PhoneNumberUtil.PhoneNumberFormat.html)
#'   for more details.
#'   
#'   `format` defaults to `"E164"`. The default can be set in option
#'   `dialr.format`.
#'   
#' @param home [ISO country code][dialr-region] for home region. If provided,
#'   numbers will be formatted for dialing from the home region.
#' @param clean Should non-numeric characters be removed? If `TRUE`, keeps
#'   numbers and leading `"+"`.
#' @param strict Should invalid phone numbers be removed? If `TRUE`, invalid
#'   phone numbers are replaced with `NA`.
#' @export
format.phone <- function(x, format = c("E164", "NATIONAL", "INTERNATIONAL", "RFC3966"),
                         home = NULL, clean = TRUE, strict = FALSE, ...) {
  if (identical(format, eval(formals()$format)))
    format <- getOption("dialr.format")
  
  format <- match.arg(format)

  validate_phone_format(format)
  validate_phone_region(home)

  phone_util <- .get_phoneNumberUtil()
  format <- .get_phone_format_from_string(format)

  out <- phone_apply(x, function(pn) {
    if (is.null(home)) {
      .jcall(phone_util, "S", "format", pn, format)
    } else {
      .jcall(phone_util, "S", "formatOutOfCountryCallingNumber", pn, home)
    }
  }, character(1))
  if (clean) out <- gsub("[^+0-9]", "", out)
  if (strict) out[!is_valid(x)] <- NA_character_

  out
}

#' @export
summary.phone <- function(object, ...) {
  out <- c(Class   = "dialr phone",
           Numbers = length(object),
           Parsed  = sum(is_parsed(object)))
  class(out) <- c("table")

  out
}

#' @rdname dialr-phone
#' @param raw If `TRUE`, the raw phone number is returned. Otherwise elements
#'   are cleaned with `format()`.
#' @export
as.character.phone <- function(x, raw = TRUE, ...) {
  if (raw) {
    x <- vapply(unclass(x), function(x) { x$raw }, character(1), USE.NAMES = FALSE)
    NextMethod()
  } else {
    as.character.default(format(x, ...))
  }
}

phone_apply <- function(x, fun, fun.value, progress = FALSE) {
  show_pb <- isTRUE(progress) && isTRUE(getOption("dialr.show_progress")) && interactive()
  if (show_pb) pb <- txtProgressBar(min = 0, max = length(x), style = 3)
  
  out <- vapply(unclass(x), function(d) {
    if (show_pb) setTxtProgressBar(pb, getTxtProgressBar(pb) + 1)
    
    if (!typeof(d$jobj) %in% "S4") return(fun.value[NA])
    fun(d$jobj)
  }, fun.value, USE.NAMES = FALSE)
  
  if (show_pb) close(pb)
  
  out
}

#' Phone number validity checks
#'
#' @description
#'
#' For each element of `x`:
#'
#' * `is_parsed(x)`: Was this successfully parsed?
#'
#' * `is_valid(x)`: Is this a valid phone number?
#'
#' * `is_possible(x)`: Is this a possible phone number? Return type depends on
#' `detailed`.
#' 
#' @details
#' 
#' Possible return values for `is_possible(x, detailed = TRUE)`:
#'
#' * `"INVALID_COUNTRY_CODE"`: The number has an invalid country calling code.
#
#' * `"INVALID_LENGTH"`: The number is longer than the shortest valid numbers
#' for this region, shorter than the longest valid numbers for this region,
#' and does not itself have a number length that matches valid numbers for
#' this region.
#'
#' * `"IS_POSSIBLE"`: The number length matches that of valid numbers for this
#' region.
#'
#' * `"IS_POSSIBLE_LOCAL_ONLY"`: The number length matches that of local numbers
#' for this region only (i.e. numbers that may be able to be dialled within an
#' area, but do not have all the information to be dialled from anywhere inside
#' or outside the country).
#'
#' * `"TOO_LONG"`: The number is longer than all valid numbers for this region.
#'
#' * `"TOO_SHORT"`: The number is shorter than all valid numbers for this
#' region.
#'
#' @section libphonenumber reference:
#'
#'   `is_valid()`: `PhoneNumberUtil.isValidNumber()`
#'
#'   `is_possible()`: `PhoneNumberUtil.isPossibleNumber()`
#'   
#'   `is_possible(detailed = TRUE)`: `PhoneNumberUtil.isPossibleNumberWithReason()`
#'   
#'   `is_possible(type = type)`: `PhoneNumberUtil.isPossibleNumberForType()`
#'   
#'   `is_possible(detailed = TRUE, type = type)`: `PhoneNumberUtil.sPossibleNumberForTypeWthReason()`
#'   
#' @param x A [phone] vector.
#' @examples
#'   x <- phone(c(0, 0123, "0412 345 678", "61412987654", "03 9123 4567", "+12015550123"), "AU")
#'
#'   is_parsed(x)
#'   is_valid(x)
#'
#'   is_possible(x)
#'   is_possible(x, detailed = TRUE)
#'
#'   is_possible(x, type = "MOBILE")
#'   is_possible(x, detailed = TRUE, type = "MOBILE")
#' @name dialr-valid
#' @family phone functions
NULL

#' @rdname dialr-valid
#' @export
is_parsed <- function(x) {
  if (!is.phone(x)) stop("`x` must be a vector of class `phone`.", call. = FALSE)
  vapply(unclass(x), function(pn) { typeof(pn$jobj) %in% "S4" }, logical(1), USE.NAMES = FALSE)
}

#' @rdname dialr-valid
#' @export
is_valid <- function(x) {
  if (!is.phone(x)) stop("`x` must be a vector of class `phone`.", call. = FALSE)
  phone_util <- .get_phoneNumberUtil()
  
  out <- phone_apply(x, function(pn) {
    .jcall(phone_util, "Z", "isValidNumber", pn)
  }, logical(1))
  out[is.na(out)] <- FALSE
  
  out
}

#' @rdname dialr-valid
#' @param detailed If `FALSE`, `is_possible` returns a logical vector. If
#'   `TRUE`, it returns a character vector with `"IS_POSSIBLE"` or the reason
#'   for failure. See Details section for possible return values.
#' @param type If provided, checks if `x` is possible for the given [phone
#'   number type][dialr-region].
#' @export
is_possible <- function(x, detailed = FALSE, type = NULL) {
  if (!is.phone(x)) stop("`x` must be a vector of class `phone`.", call. = FALSE)
  validate_phone_type(type)
  
  phone_util <- .get_phoneNumberUtil()
  
  if (is.null(type)) {
    if (detailed) {
      out <- phone_apply(x, function(pn) {
        .jstrVal(.jcall(phone_util,
                        "Lcom/google/i18n/phonenumbers/PhoneNumberUtil$ValidationResult;",
                        "isPossibleNumberWithReason",
                        pn))
      }, character(1))
    } else {
      out <- phone_apply(x, function(pn) {
        .jcall(phone_util, "Z", "isPossibleNumber", pn)
      }, logical(1))
    }
  } else {
    type <- .get_phone_type_from_string(type)
    if (detailed) {
      out <- phone_apply(x, function(pn) {
        .jstrVal(.jcall(phone_util,
                        "Lcom/google/i18n/phonenumbers/PhoneNumberUtil$ValidationResult;",
                        "isPossibleNumberForTypeWithReason",
                        pn, type))
      }, character(1))
    } else {
      out <- phone_apply(x, function(pn) {
        .jcall(phone_util, "Z", "isPossibleNumberForType", pn, type)
      }, logical(1))
    }
  }
    
  if (!detailed) out[is.na(out)] <- FALSE
  
  out
}
